sw/device/lib/base/mmio.h

```diff
@@ -5,6 +5,12 @@
 #ifndef SW_DEVICE_LIB_BASE_MMIO_H_
 #define SW_DEVICE_LIB_BASE_MMIO_H_
 
+// This file is included in C and C++, and, as such, needs to be marked as
+// extern "C" in C++ to make sure linking works out.
+#ifdef __cplusplus
```

Comment:
#1300 doesn't have this and I didn't run into any linking issues. Probably will be apparent to me soon and I wonder why you need this here.

Reply:
If you pull this into a `.c` file, the function symbols will be what you expect: `reg32_read`, and so on. If you pull it into a `.cc` file, and you don't mark them as `extern "C"`, they will instead look like `_ZN10reg32_readEblahblahblah`, because C++ includes the argument types in the mangled symbol name. Since DIFs are implemented in a `.c` file, but these functions are also referred to by `mock_mmio.cc`, they need to be marked as `extern "C"` to disable name mangling.

Reply:
Thanks for the explanation, and apologies for not being clear. I was wondering why you need `extern "C"` in the file itself as apposed to the point where it is included in the `.cc` file.

Reply:
Please correct me if I am wrong, but I think this is a fairly standard approach, It avoids the need to have the guard in every .cc file where this header is used.

Reply:
Yeah, it's pretty standard. But I felt it was prudent to point out that the file was, unlike most OT headers, C and C++ safe.